tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/ScalaJSClassEmitter.scala

```diff
@@ -72,6 +74,73 @@ private[scalajs] final class ScalaJSClassEmitter(
 
   private[emitter] def semantics: Semantics = linkingUnit.semantics
 
+  private[emitter] lazy val typeTags: Map[LinkedClass, Int] = {
+    import Definitions.ReservedTags
+
+    val childrenMap =
+      linkingUnit.classDefs.groupBy(_.superClass.map(_.name)).withDefaultValue(Nil)
+
+    // Some classes may have parents that have been removed
+    val orphans = for(cls <- linkingUnit.classDefs;
+                    parent <- cls.superClass
+                    if !linkedClassByName.isDefinedAt(parent.name)) yield cls
+
+    val roots = childrenMap(None) ++ orphans
+
+    @tailrec def getTypeTags(curTag: Int, stack: Seq[LinkedClass],
+      accu: Map[LinkedClass, Int] = Map.empty): Map[LinkedClass, Int] = stack match {
+      case cur +: rest if ReservedTags.isDefinedAt(cur.name.name) =>
+        val children = childrenMap(Some(cur.name.name))
+        val tag = ReservedTags(cur.name.name)
+        getTypeTags(curTag, children ++ rest, accu + (cur -> tag))
+
+      case cur +: rest =>
+        val nextTag = curTag + 1
+        val children = childrenMap(Some(cur.name.name))
+        getTypeTags(nextTag, children ++ rest, accu + (cur -> curTag))
+
+      case Nil =>
+        // Sanity check: all classDefs must have a tag
+        assert(linkingUnit.classDefs.forall(accu.isDefinedAt))
+        accu
+    }
+
+    // The smallest tags are reserved
+    getTypeTags(ReservedTags.values.max + 1, roots)
+  }
+
+  private[emitter] lazy val subtypeTags: Map[String, List[Int]] = {
```

Comment:
This is only used in `subtypeIntervals`. Its computation should be moved as an inner `val` or `def` within `subtypeIntervals`.